### PR TITLE
Add a --nightly | -n flag to mach run commands for linux

### PR DESCRIPTION
**Pull request by o0Ignition0o**
First tries to download and extract a specific nightly version to run mach commands against.

I chose to split the Pull requests for each platform to avoid submitting a huge one, and to make sure I get the logic right.
I'm able to download / extract a nightly version, and I keep nightly versions in the target folder.
Windows and Mac OS support will be filed in separate PRs.
This is part of step two for #19505 

The mentor on the issue is jdm 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because it is part of a ./mach command.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19947)

### CODE REVIEW

File: python/servo/command_base.py [hunk 1]
```diff
@@ -395,6 +399,79 @@ def get_binary_path(self, release, dev, android=False):
...
+        if not nightly_date:
+            print(
+                "No nightly date has been provided although the --nightly or -n flag has been passed.")
+            sys.exit(1)
+        # Will alow us to fetch the relevant builds from the nightly repository
+        os_prefix = "linux"
+        if is_windows():
+            print("The nightly flag is not supported on windows yet.")
+            sys.exit(1)
+            os_prefix = "windows-msvc"
```
Comment (tigercosmos) [inline python/servo/command_base.py]: why we change `os_prefix` after the program has exited?

Comment (o0Ignition0o) [reply python/servo/command_base.py]: Good catch ! I was actually working on a multi platform solution, and then decided to split the merge requests. I'll remove it :)

**Comment by jdm** [comment]
r? @jdm

Sorry about the delay; let's get a more appropriate reviewer on this.

**Comment by o0Ignition0o** [comment]
Hi :)
No Problem, i've started working on the windows support since then. I'll put it in a separate PR :)

**Comment by o0Ignition0o** [comment]
``` It looks like rustup is not installed. See instructions at https://github.com/servo/servo/#setting-up-your-environment
Build FAILED in 0:00:00
```

This looks weird, but I'm not really sure it's related to the changes I've made, any thoughts ?

**Comment by jdm** [comment]
You can ignore the taskcluster failure.

**Comment by o0Ignition0o** [comment]
Ok thanks !
I guess it should be reviewable then :)

**Comment by o0Ignition0o** [comment]
Hey there :) is anyone available for some suggestions or a code review ? I'm not at home but I'll ask on IRC as soon as I come home :)

**Comment by jdm** [comment]
Ugh, sorry for forgetting about this. I'll review it today.

File: python/servo/post_build_commands.py [hunk 2]
```diff
@@ -160,14 +162,17 @@ def run(self, params, release=False, dev=False, android=None, debug=False, debug
...
         'params', nargs='...',
         help="Command-line arguments to be passed through to Servo")
-    def rr_record(self, release=False, dev=False, bin=None, params=[]):
+    def rr_record(self, release=False, dev=False, bin=None, nightly=None, params=[]):
         env = self.build_env()
         env["RUST_BACKTRACE"] = "1"
 
-        servo_cmd = [bin or self.get_binary_path(release, dev)] + params
+        servo_cmd = [bin or self.get_nightly_binary_path(
+            nightly.strip()) or self.get_binary_path(release, dev)] + params
```
Comment (jdm) [inline python/servo/post_build_commands.py]: nit: let's join these lines, or split after the `or` instead.

File: python/servo/command_base.py [hunk 3]
```diff
@@ -388,6 +392,77 @@ def get_binary_path(self, release, dev, android=False):
...
+            tree = XML(response)
+            namespaces = {'ns': tree.tag[1:tree.tag.index('}')]}
+            file_to_download = tree.find('ns:Contents', namespaces).find(
+                'ns:Key', namespaces).text
+        except urllib2.URLError as e:
+            print("Could not fetch the available nightly versions from the repository : {}".format(
+                e.reason))
+            sys.exit(1)
+        except AttributeError as e:
+            print("Could not fetch a nightly version for date {} and platform {} ".format(
```
Comment (jdm) [inline python/servo/command_base.py]: nit: remove the extra space after {}

File: python/servo/command_base.py [hunk 4]
```diff
@@ -388,6 +392,77 @@ def get_binary_path(self, release, dev, android=False):
...
+                nightly_date, os_prefix))
+            sys.exit(1)
+
+        nightly_target_directory = path.join(self.context.topdir, "target")
+        # Once extracted, the nightly folder name is the tar name without the extension
+        # (eg /foo/bar/baz.tar.gz extracts to /foo/bar/baz)
+        destination_file = path.join(
+            nightly_target_directory, file_to_download)
+        destination_folder = destination_file[:destination_file.index('.')]
```
Comment (jdm) [inline python/servo/command_base.py]: We can use `os.path.splitext(destination_file)[0]` instead.

**Review by jdm — CHANGES_REQUESTED** [review]
This looks really great! Thanks!

File: python/servo/command_base.py (hunk 4 above)
Comment (o0Ignition0o) [reply python/servo/command_base.py]: I didn't know about this method, pretty cool !

**Comment by o0Ignition0o** [comment]
This looks better now I guess :)

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
```
Error running mach:

    ['run', '-r', '-o', 'output.png']

The error occurred in the implementation of the invoked mach command.

This should never occur and is likely a bug in the implementation of that
command. Consider filing a bug for this issue.

If filing a bug, please include the full output of mach, including this error
message.

The details of the failure are as follows:

AttributeError: 'NoneType' object has no attribute 'strip'

  File "/Users/servo/buildbot/slave/mac-rel-wpt1/build/python/servo/post_build_commands.py", line 100, in run
    args = [bin or self.get_nightly_binary_path(nightly.strip()) or self.get_binary_path(release, dev)]
```

**Comment by o0Ignition0o** [comment]
Ok I found two ways to handle this issue: 
I can use an if / else construct to check if the string is empty or not.
I can change the command argument to default to "", but that would break the "command arguments consistency" eg : 

`  @CommandArgument('--debugger', default=None, type=str,
                     help='Name of debugger to use.')
    @CommandArgument('--headless', '-z', action='store_true',
                     help='Launch in headless mode')
    @CommandArgument('--software', '-s', action='store_true',
                     help='Launch with software rendering')
    @CommandArgument('--bin', default=None,
                     help='Launch with specific binary')
    @CommandArgument('--nightly', '-n', default="",
                     help='Specify a YYYY-MM-DD nightly build to run')
    @CommandArgument(
        'params', nargs='...',
        help="Command-line arguments to be passed through to Servo")
    def run(self, params, release=False, dev=False, android=None, debug=False, debugger=None,
            headless=False, software=False, bin=None, nightly=""):  `

I think using "" as default would be more elegant, but I can understand that CommandArgument / public api consistency might be more important.
Which one would you rather have?

**Comment by o0Ignition0o** [comment]
Ok I went for the empty default string, please let me know if you think the other approach is better :)

**Comment by jdm** [comment]
@o0Ignition0o Another option would be to strip the value inside get_nightly_binary_path once it has been checked for None.

**Comment by o0Ignition0o** [comment]
It is indeed a better approach !

**Pull request merged**